scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CipherSuite.java

```diff
@@ -50,16 +54,28 @@
-	TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256(0xC023, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_CBC, MACAlgorithm.HMAC_SHA256),
-	TLS_PSK_WITH_AES_128_CCM_8(0xC0A8, KeyExchangeAlgorithm.PSK, Cipher.AES_128_CCM_8, MACAlgorithm.NULL),
-	TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8(0xC0AE, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_CCM_8, MACAlgorithm.NULL);
+	TLS_NULL_WITH_NULL_NULL(0x0000, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.NULL, Cipher.NULL, MACAlgorithm.NULL, true),
+	TLS_PSK_WITH_AES_128_CCM_8(0xC0A8, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_128_CCM_8, MACAlgorithm.NULL, false),
+	TLS_PSK_WITH_AES_128_CBC_SHA256(0x00AE, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_128_CBC, MACAlgorithm.HMAC_SHA256, false),
+	TLS_PSK_WITH_AES_256_CCM_8(0xC0A9, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_256_CCM_8, MACAlgorithm.NULL, true),
+	TLS_PSK_WITH_AES_128_CCM(0xC0A4, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_128_CCM, MACAlgorithm.NULL, true),
+	TLS_PSK_WITH_AES_256_CCM(0xC0A5, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_256_CCM, MACAlgorithm.NULL, true),
+	TLS_PSK_WITH_AES_128_GCM_SHA256(0x00A8, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.PSK, Cipher.AES_128_GCM, MACAlgorithm.NULL, true),
 
+	/**See <a href="https://tools.ietf.org/html/rfc5489#section-3.2">RFC 5489</a> for details*/
+	TLS_ECDHE_PSK_WITH_AES_128_CBC_SHA256(0xC037, CertificateKeyAlgorithm.NONE, KeyExchangeAlgorithm.ECDHE_PSK, Cipher.AES_128_CBC, MACAlgorithm.HMAC_SHA256, true),
+
+	TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8(0xC0AE, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_CCM_8, MACAlgorithm.NULL, false),
+	TLS_ECDHE_ECDSA_WITH_AES_256_CCM_8(0xC0AF, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_256_CCM_8, MACAlgorithm.NULL, true),
+	TLS_ECDHE_ECDSA_WITH_AES_128_CCM(0xC0AC, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_CCM, MACAlgorithm.NULL, true),
+	TLS_ECDHE_ECDSA_WITH_AES_256_CCM(0xC0AD, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_256_CCM, MACAlgorithm.NULL, true),
+	TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256(0xC023, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_CBC, MACAlgorithm.HMAC_SHA256, false),
+	TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384(0xC024, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_256_CBC, MACAlgorithm.HMAC_SHA384, true, PRFAlgorithm.TLS_PRF_SHA384),
+	TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256(0xc02b, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_128_GCM, MACAlgorithm.NULL, true),
+	TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA(0xC00A, CertificateKeyAlgorithm.EC, KeyExchangeAlgorithm.EC_DIFFIE_HELLMAN, Cipher.AES_256_CBC, MACAlgorithm.HMAC_SHA1, true),
+	;
```

Comment:
extended := new

My idea was to provide backwards compatibility and small CLIENT_HELLOs by default.

Reply:
The order was just my idea about the default order.

Reply:
> extended := new

new comparing to what ? 
If this is compared to previous commit, TLS_ECDHE_PSK_WITH_AES_128_CBC_SHA256 was present and should be added too ?
I don't know if this backward compatibility is really needed (by the way, I believe this PR break the cipher suite list order anyway) ?
I mean it would make more sense to me if we make the difference between recommended and obsolete cipher ? Or maybe, non-extended could just be the COAP recommended/mandatory : 
```
TLS_PSK_WITH_AES_128_CCM_8
TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8
```

> The order was just my idea about the default order.

I believe the idea should be to put more  secure in first ?
ECDHE first ?
GCM >= CCM > CBC ? https://en.wikipedia.org/wiki/Transport_Layer_Security#Cipher

Not directly linked but do you know if we support rfc7366 ? especially for AES-128-CBC-SHA256 (https://github.com/OpenMobileAlliance/OMA_LwM2M_for_Developers/issues/6#issuecomment-244017113)

Reply:
> If this is compared to previous commit, TLS_ECDHE_PSK_WITH_AES_128_CBC_SHA256 was present and should be added too ?

My intention was to "new compared to 1.0." (but more assuming that than validating that :-) ).

The default order will be (defined in determineCipherSuitesFromConfig):
TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
TLS_PSK_WITH_AES_128_CCM_8
TLS_PSK_WITH_AES_128_CBC_SHA256
if both, certificate based and psk based suites are used. I think, this is the same as for 1.0.

For the new ones I don't see, that the default order is significant. I see, that everyone should then be able to chose the best order on his case.

Reply:
> rfc7366

I don't know. But this PR doesn't change it :-).

Reply:
> I don't know if this backward compatibility is really needed (by the way, I believe this PR break the cipher suite list order anyway) ?

My idea was, that for the client side it may be a good idea, not to have a larger CLIENT_HELLO by default. For the server the "extended" setup could be used, if "role exchange" is not used.

Reply:
so `//Cipher suites, ordered by default preference` in CipherSuite is not consistent with determineCipherSuitesFromConfig.

I'm still confused with `extendedCipherSuites` concept. If we keep it a concept like this, I will just go to separate mandatory cipher and optional one.

About the order, I'm pretty sure that the good practice is [more secure first](https://github.com/ssllabs/research/wiki/SSL-and-TLS-Deployment-Best-Practices#24-select-best-cipher-suites).

Reply:
Unfortunately,  `mandatory` for RFC7252 seems to be
TLS_PSK_WITH_AES_128_CCM_8 and TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8.
I didn't want to remove the CBC support from the "default"

Reply:
> so //Cipher suites, ordered by default preference in CipherSuite is not consistent with determineCipherSuitesFromConfig.

Good catch.

Reply:
> I didn't want to remove the CBC support from the "default"

Any reason ?

Reply:
> About the order, I'm pretty sure that the good practice is more secure first.

Let' merge it and you propose a PR with the best practice.

Reply:
> Any reason ?

Backwards compatibility and LWM2M 1.0 support.

Reply:
May be worth to mention:
I added the suites to be able to test with the go-DTLS implementation. 
Afterwards I was asked to add some more cipher suites (issue #884).
But I didn't want to clean up more. I got also aware of  "Encrypt then MAC" (RFC 7366), but I didn't have the time to check it.

Reply:
We will release a 2.0.0, if we think this "break" make sense, it's the right moment.
About LWM2M 1.0 support, this is more a Leshan default cipher question.

Here we should talk about CoAP meaning, and for me :
 - either by default we put all the cipher available
- or we have a kind of good reason to put some as default rather than other (like CoAP mandatory vs non mandatory/ recommended vs not recommended like CBC)

Do you see what I mean ?

Reply:
I can do the PR if we succeed to agree about what we should do :)

Reply:
> all the cipher available

That may be good for now, but if more cipher suites added it will end up similar to other libraries, where you need to select the cipher suites or you get a large CLIENT_HELLO.

My proposal:
- create a "voting issue" about the wanted default cipher suites for 2.0 
- fix the comment (me)
- PR for the new order 8you)
- when the voting is finished in a couple of days, adjust the default.

Reply:
Read with 1 and 2.
Merge now to prepare for 3?

Reply:
yep